freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/KeytabCleanupService.java

```diff
@@ -0,0 +1,155 @@
+import com.sequenceiq.freeipa.entity.Stack;
+import com.sequenceiq.freeipa.kerberosmgmt.exception.DeleteException;
+import com.sequenceiq.freeipa.service.freeipa.FreeIpaClientFactory;
+import com.sequenceiq.freeipa.service.freeipa.host.HostDeletionService;
+
+@Service
+public class KeytabCleanupService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(KeytabCleanupService.class);
+
+    private static final String SERVICE_PRINCIPAL_DELETION_FAILED = "Failed to delete service principal.";
+
+    @Inject
+    private FreeIpaClientFactory freeIpaClientFactory;
+
+    @Inject
+    private KerberosMgmtRoleComponent roleComponent;
+
+    @Inject
+    private KerberosMgmtVaultComponent vaultComponent;
+
+    @Inject
+    private HostDeletionService hostDeletionService;
+
+    @Inject
+    private KeytabCacheService keytabCacheService;
+
+    @Inject
+    private KeytabCommonService keytabCommonService;
+
+    public void deleteServicePrincipal(ServicePrincipalRequest request, String accountId) throws FreeIpaClientException, DeleteException {
+        LOGGER.debug("Request to delete service principal for account {}: {}", accountId, request);
+        Stack freeIpaStack = keytabCommonService.getFreeIpaStack(request.getEnvironmentCrn(), accountId);
+        String realm = keytabCommonService.getRealm(freeIpaStack);
+        String canonicalPrincipal = keytabCommonService.constructPrincipal(request.getServiceName(), request.getServerHostName(), realm);
+        FreeIpaClient ipaClient = freeIpaClientFactory.getFreeIpaClientForStack(freeIpaStack);
+        delService(canonicalPrincipal, ipaClient);
+        VaultPathBuilder vaultPathBuilder = new VaultPathBuilder()
+                .withSecretType(VaultPathBuilder.SecretType.SERVICE_KEYTAB)
+                .withAccountId(accountId)
+                .withEnvironmentCrn(request.getEnvironmentCrn())
+                .withClusterCrn(request.getClusterCrn())
+                .withServerHostName(request.getServerHostName())
+                .withServiceName(request.getServiceName());
+        vaultComponent.recursivelyCleanupVault(vaultPathBuilder.withSubType(VaultPathBuilder.SecretSubType.SERVICE_PRINCIPAL).build());
+        vaultComponent.recursivelyCleanupVault(vaultPathBuilder.withSubType(VaultPathBuilder.SecretSubType.KEYTAB).build());
+        roleComponent.deleteRoleIfItIsNoLongerUsed(request.getRoleName(), ipaClient);
+        keytabCacheService.deleteByEnvironmentCrnAndPrincipal(request.getEnvironmentCrn(), canonicalPrincipal);
+    }
+
+    public void deleteHost(HostRequest request, String accountId) throws FreeIpaClientException {
+        FreeIpaClient ipaClient = freeIpaClientFactory.getFreeIpaClientByAccountAndEnvironment(request.getEnvironmentCrn(), accountId);
+        cleanupServicesForHost(request, accountId, ipaClient);
+        hostDeletionService.deleteHostsWithDeleteException(ipaClient, Set.of(request.getServerHostName()));
```

Comment:
any freeipa call can fail for any reason, most likely because the connection breaks for some reason. Anyway `DeleteException` is just another wrapper
roles and vault is not that critical if they hang around